File: problem-solving/adventofcode2020/day5.py
# https://adventofcode.com/2020/day/5

input_file = open('./day5.txt', 'r')
input_list = input_file.read().splitlines()

def get_highest_seat_id(input_list):
    highest_seat_id = 0
    for input in input_list:
        row_string, col_string = input[0:7], input[7:10]
        row = compute_row(row_string)
        col = compute_col(col_string)
        unique_seat_id = get_unique_seat_id(row, col)
        highest_seat_id = max(highest_seat_id, unique_seat_id)
    return highest_seat_id

def find_my_missing_seat(input_list):
    seat_ids = []
    for input in input_list:
        row_string, col_string = input[0:7], input[7:10]
        row = compute_row(row_string)
        col = compute_col(col_string)
        unique_seat_id = get_unique_seat_id(row, col)
        seat_ids.append(unique_seat_id)
    seat_ids.sort()
    my_id = 0
    for i in range(1, len(seat_ids)):
        if seat_ids[i] - seat_ids[i-1] != 1:
            my_id = seat_ids[i]-1
            break
    return my_id

def compute_row(row_string):
    start = 0
    end = 127
    for v in list(row_string):
        diff = (end-start)//2+1
        if v == "F":
            end -= diff
        elif v == "B":
            start += diff
    return start

def compute_col(col_string):
    start = 0
    end = 7
    for v in list(col_string):
        diff = (end-start)//2+1
        if v == "L":
            end -= diff
        elif v == "R":
            start += diff
    return start

def get_unique_seat_id(row, col):
    return row * 8 + col

if __name__ == "__main__":
    print(get_highest_seat_id(input_list))
    print(find_my_missing_seat(input_list))

File: problem-solving/adventofcode2020/day5.cpp
// https://adventofcode.com/2020/day/5
#include <iostream>
#include <string>
#include <vector>
#include <fstream>
#include <algorithm>

using namespace std;

vector<string> readInput(string filepath)
{
    vector<string> inputList;
    ifstream inputFile;
    inputFile.open(filepath);

    string line;
    while (inputFile >> line)
    {
        inputList.push_back(line);
    }
    inputFile.close();
    return inputList;
}

int computeRow(string &rowStr)
{
    int start = 0;
    int end = 127;
    for (char &s : rowStr)
    {
        int diff = (end - start) / 2 + 1;
        if (s == 'F')
        {
            end -= diff;
        }
        else if (s == 'B')
        {
            start += diff;
        }
    }
    return start;
}

int computeCol(string &colStr)
{
    int start = 0;
    int end = 7;
    for (char &s : colStr)
    {
        int diff = (end - start) / 2 + 1;
        if (s == 'L')
        {
            end -= diff;
        }
        else if (s == 'R')
        {
            start += diff;
        }
    }
    return start;
}

int getUniqueSeatID(int row, int col)
{
    return row * 8 + col;
}

int getHighestSeatID(const vector<string> &inputList)
{
    int highestSeatID = 0;
    for (int i = 0; i < inputList.size(); i++)
    {
        string rowStr = inputList[i].substr(0, 7);
        string colStr = inputList[i].substr(7, 10);
        int row = computeRow(rowStr);
        int col = computeCol(colStr);
        int uniqueSeatID = getUniqueSeatID(row, col);
        if (highestSeatID < uniqueSeatID)
        {
            highestSeatID = uniqueSeatID;
        }
    }
    return highestSeatID;
}

int findMyMissingSeat(const vector<string> &inputList)
{
    vector<int> seatIDs;
    for (int i = 0; i < inputList.size(); i++)
    {
        string rowStr = inputList[i].substr(0, 7);
        string colStr = inputList[i].substr(7, 10);
        int row = computeRow(rowStr);
        int col = computeCol(colStr);
        int uniqueSeatID = getUniqueSeatID(row, col);
        seatIDs.push_back(uniqueSeatID);
    }

    sort(seatIDs.begin(), seatIDs.end());
    int myID = 0;
    for (int i = 1; i < seatIDs.size(); i++)
    {
        if (seatIDs[i] - seatIDs[i - 1] != 1)
        {
            myID = seatIDs[i] - 1;
            break;
        }
    }
    return myID;
}

int solvePart1(const vector<string> &inputList)
{
    return getHighestSeatID(inputList);
}

int solvePart2(const vector<string> &inputList)
{
    return findMyMissingSeat(inputList);
}

int main()
{
    vector<string> inputList = readInput("./day5.txt");
    cout << solvePart1(inputList) << endl;
    cout << solvePart2(inputList) << endl;
}